### ISSUE TITLE
About the val directory

### ISSUE THREAD

**Issue by mnc1423** [perslev/MultiPlanarUNet#1]
i was wondering what the val directory is and what you fill it with 
giving me an error that says no image in val 

can i use the same data in the training files?

**Comment by erikbdam** [comment]
Hi MNC,

The validation directory is used following the standard Machine Learning setup. You have a training collection for optimizing the parameters (the images in your training folder) and then you have validation collection for estimating the performance on unseen data and possibly optimizing hyper-parameters. 

While debugging, you can use the same data as the training files to avoid this error. But as soon as you start looking at the performance, the training and validation files should be different. 

Cheers,
Erik

**Comment by perslev** [comment]
Also, note that you can perform training without validation using the --no_val flag. I.e.:

mp train --no_val

**Comment by mnc1423** [comment]
Hey Erik
Thanks for the reply! Also was wondering how many datasets is needed. i'm currently using 30 CT pics and should this be enough?

Thanks again :)

**Comment by erikbdam** [comment]
Well, that is impossible to answer with no knowledge of your data. The required amount of training data depends on how challenging the problem is. Are the scans noisy or clean? Is there a clear boundary delineating the desired object(s)? What is the biological variation in the anatomy? Is there variation in the way the scan is done (positioning, settings, scanner)? The harder the problem is, the more training data you need. 

My suggestion would be to split fifty-fifty while you play a bit with the method. And then when you are ready to make a proper validation, you should consider a five-fold cross-validation (allowing you to train on 25 scans instead of the 15 given by a fifty-fifty split).

**Comment by mnc1423** [comment]
Sorry about the lack of information. Using Kits19 DataSet and gonna test on other real CTs 
I will follow you advice and start from there :)
Thank you for the help

**Comment by mnc1423** [comment]
Aside from that i'm getting OOM when train_fusion i tried decreasing the images_per_round to 2 but still seem to be getting it. 
Currently using TITAN XP, don't see an option for decreasing the batch size 
Anything i can do?
Thx in advance

**Comment by perslev** [comment]
Hi, and thank you for your feedback.

Could you please provide the following information:

- Are you getting a TensorFlow memory error? Otherwise, I suspect you may be running out of RAM. The fusion training process is quite memory consuming when processing large images. Please see below.
- How much RAM does your system have?
- How big are the sampled images? Please refer to the 'dim' parameter under 'build' in the train_hparams.yaml file, or in the training log files.

The train_fusion script is quite primitive at this point. I am currently re-writing it to help prevent issues like those you are experiencing. For now, I suggest you either reduce the images_per_round parameter to 1 and see if that fixes your issue. Alternatively, you may skip fusion training entirely by invoking the prediction script with the --sum_fusion flag (note that this will reduce performance). E.g.:

mp predict --sum_fusion

The memory usage of predicting without fusion is, however, roughly equal to running train_fusion with 1 image per round.

Cheers,
Mathias

**Comment by mnc1423** [comment]
Hey perslev! Thank you for the reply
1. Yes currently getting a TensorFLow memory error
2. I'm currently using a server that has 128GB of DDR4 ram 
3. 
build: &BUILD
  
  Hyperparameters passed to the Model.build and __init__ methods
  
  model_class_name: "UNet"
  n_classes: 2
  n_channels: 1
  dim: 512
  complexity_factor: 2
  out_activation: "softmax"
  l1_reg: False
  l2_reg: False
  biased_output_layer: True
  depth: 4

Thank you hope this helps

**Comment by perslev** [comment]
Hi,

Thanks! I assume you were able to run training successfully on this system? The problem in that case, as you mentioned, could be that the fusion model although small in size is trained with a very large batch size, and is placed on the same GPU next to the UNet.

Please pull the latest repository and checkout on branch 'v.0.2.2' and try re-running the fusion script. It now accepts a --batch_size parameter, which you can decrease until you no longer experience the error. The default value for batch_size has also been decreased significantly, so you may not have to manually set the parameter at all.

These changes will be merged with the master branch at a later point when the script has been re-written altogether.

Cheers,
Mathias

**Comment by mnc1423** [comment]
Thanks Mathias,

Yes i was able to run the script successfully with images_per_round set to 1
i will try out the new repository and check out the performance
Thanks for the quick update and reply, Awesome script btw

Cheers
Mnc

**Comment by perslev** [comment]
You are welcome! If you experience any other issues or have further questions, please let us know.
